src/psyclone/psyGen.py

```diff
@@ -6598,7 +6627,32 @@ def nodes_to_code_block(parent, statements):
         if not statements:
             return None
 
-        code_block = CodeBlock(statements, parent=parent)
+        # Determine whether this code block is a statement or an
+        # expression.
+        if parent.children:
+            # Find the previous entry in the PSyIR
+            prev_node = parent.children[-1]
+            if isinstance(
+                    prev_node, (Directive, GlobalSum, HaloExchange, Kern, Loop,
+                                IfBlock, Assignment, Return)):
+                # The previous node is a statement, therefore this
+                # code block must be a statement.
+                structure = CodeBlock.Structure.STATEMENT
+            elif isinstance(prev_node, (Reference, Operation, Literal)):
+                # The previous node is an expression, therefore this
+                # code block must be an expression.
+                structure = CodeBlock.Structure.EXPRESSION
+            else:
+                raise InternalError(
+                    "Fparser2ASTProcessor:nodes_to_codeblock: Unsupported "
+                    "node type '{0}' found.".format(type(prev_node).__name__))
+        else:
```

Comment:
OK until here, but the else condition (Codeblock has no sibling yet) does not imply it is an statement, it can be the first node of an expression,

See my example on the unit-tests below.

Reply:
Yes, quite right, the code would never work. I've updated taking your advice i.e. checking whether the parent is a schedule or not (If yes then it's a statement, if not it's an expression), raising an exception if it is a directive (as these require further discussion in my opinion).